huge number of "replacement function 'operator new' cannot be declared inline" warnings

`MOZALLOC_INLINE` is applied to `operator new`, which is apparently not legal and ignored.  This causes a warning in a large percentage of source files, which is an astronomical amount of output:

```
120:01.56 Warning: -Winline-new-delete in /Users/pale/jenkins/workspace/palemoon/obj-x86_64-apple-darwin12.5.0/dist/include/mozilla/mozalloc.h: replacement function 'operator new' cannot be declared 'inline'
120:01.56 ../../dist/include/mozilla/mozalloc.h:204:21: warning: replacement function 'operator new' cannot be declared 'inline' [-Winline-new-delete]
120:01.64 MOZALLOC_EXPORT_NEW MOZALLOC_INLINE
120:01.69                     ^
120:01.84 ../../dist/include/mozilla/mozalloc.h:44:27: note: expanded from macro 'MOZALLOC_INLINE'
120:01.90 #  define MOZALLOC_INLINE MOZ_ALWAYS_INLINE_EVEN_DEBUG
120:01.96                           ^

120:02.16 ../../dist/include/mozilla/Attributes.h:42:75: note: expanded from macro 'MOZ_ALWAYS_INLINE_EVEN_DEBUG'
120:02.31 #  define MOZ_ALWAYS_INLINE_EVEN_DEBUG     __attribute__((always_inline)) MOZ_INLINE
120:02.47                                                                           ^
120:02.63 ../../dist/include/mozilla/Attributes.h:20:33: note: expanded from macro 'MOZ_INLINE'
120:02.71 #  define MOZ_INLINE            inline
120:02.77                                 ^
```

Compiler is

```
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 5.1 (clang-503.0.40) (based on LLVM 3.4svn)
Target: x86_64-apple-darwin12.5.0
Thread model: posix
```

--- a/configure.in
+++ b/configure.in
@@ -1426,8 +1426,12 @@ if test "$GNU_CXX"; then
 
     # Turn off the following warnings that -Wall turns on:
     # -Wno-invalid-offsetof - we use offsetof on non-POD types frequently
+    # -Wno-inline-new-delete - we inline 'new' and 'delete' in mozalloc
+    #   for performance reasons, and because GCC and clang accept it (though
+    #   clang warns about it).
     #
     MOZ_CXX_SUPPORTS_WARNING(-Wno-, invalid-offsetof, ac_cxx_has_wno_invalid_offsetof)
+    MOZ_CXX_SUPPORTS_WARNING(-Wno-, inline-new-delete, ac_cxx_has_wno_inline_new_delete)
 
     if test -z "$INTEL_CXX" -a -z "$CLANG_CXX"; then
        # Don't use -Wcast-align with ICC or clang

--- a/js/src/configure.in
+++ b/js/src/configure.in
@@ -1167,8 +1167,11 @@ if test "$GNU_CC"; then
 
     # Turn off the following warnings that -Wall turns on:
     # -Wno-unused - lots of violations in third-party code
+    # -Wno-inline-new-delete - we inline 'new' and 'delete' in mozalloc
     #
     _WARNINGS_CFLAGS="${_WARNINGS_CFLAGS} -Wno-unused"
+
+    MOZ_CXX_SUPPORTS_WARNING(-Wno-, inline-new-delete, ac_cxx_has_wno_inline_new_delete)
 
     if test -z "$INTEL_CC" -a -z "$CLANG_CC"; then
        # Don't use -Wcast-align with ICC or clang
@@ -1665,6 +1668,10 @@ ia64*-hpux*)
         # it off on MSVC, too.
         CFLAGS="$CFLAGS -wd4244"
         CXXFLAGS="$CXXFLAGS -wd4244 -wd4251"
+        if test -n "$CLANG_CL"; then
+            # Suppress the clang-cl warning for the inline 'new' and 'delete' in mozalloc
+            CXXFLAGS="$CXXFLAGS -Wno-inline-new-delete"
+        fi
         # make 'foo == bar;' error out
         CFLAGS="$CFLAGS -we4553"
         CXXFLAGS="$CXXFLAGS -we4553"

